dhis2/dhis2-core
# fix: organisation unit special root based queries as sub-paths [DHIS2-13901]

### Summary
Removes special parameters `includeChildren`, `includeDescendants` and `includeAncestors` as well as `level` from endpoint `/api/organisationUnits/{uid}`.

Instead these are used as dedicated paths (endpoints):
* `/api/organisationUnits/{uid}/children` (direct children)
* `/api/organisationUnits/{uid}/children?level=2` (children 2 levels down, that means all children of the direct children)
* `/api/organisationUnits/{uid}/descendants`
* `/api/organisationUnits/{uid}/ancestors`

This is **not** merely a change from a request parameter to a path. When this was solved as request parameter this was based on single object lookup with some special list lookup on top. This meant the result handling of the list items was the handling intended for single objects. Whereas the new path solution is using the same handling as the normal list `/api/organisationUnits` has. This is most notably in the default fields included which is just `id` and `displayName`  for a list whereas it is most of the properties for a single object.

### Cleanup
The special list for single object in the OU controller was the only reason the `getEntity` method was returning a list, not a single entity. With that now gone the method could change to just return the expected single object value.
In case the value does not exist a `NotFoundException` is thrown. This allowed to remove that check and throw code at most call sites. 

This is now also consistent with the existing `/api/organisationUnits/{uid}/parents` list which now is based on the same `getEntityList` implementation with special loading function.


### Compatibility
* With this behaviour propagating to all callers of `getEntity` this meant 2 callers for an update of a `User` would now throw `NotFoundException` instead of `ConflictException` when trying to update a user that does not exist.
* Special list now have same rendering as the usual list `/api/organisationUnits`. This means without the `fields` parameter they contain less properties then their former parameter counterparts would have returned.
* Special lists can now be used with `fields` and `filter` 

### Manual Testing
Use the 4 new API path described above and check they match the expectations. For example, check against the organisation tree shown in an app.

* `/api/organisationUnits/{uid}/children` (only direct children)
* `/api/organisationUnits/{uid}/children?level={n}` (children n level down, where 1 is direct children, 0 would the the unit itself )
* `/api/organisationUnits/{uid}/descendants` (the unit and all its children and grant-children - any depth)
* `/api/organisationUnits/{uid}/ancestors` (the unit and all its parents up to the root)
* `/api/organisationUnits/{uid}/parents` (only the parents of the unit but not the unit itself)

For example

* `/api/organisationUnits/Rp268JB6Ne4/descendants?fields=id,name,level`
* `/api/organisationUnits/Rp268JB6Ne4/parents?fields=id,name,level`
* `/api/organisationUnits/Rp268JB6Ne4/ancestors?fields=id,name,level`
* `/api/organisationUnits/ImspTQPwCqd/children?level=1`

### Automatic Testing
Added new controller test scenarios for the new endpoints.
